### DISCUSSION TITLE
Не могу изменить активный цвет текста в компоненте TabsItem

### DISCUSSION THREAD

**Discussion by Rezetchik** [VKCOM/VKUI#6055] [Q&A] [answered]
Вопрос по поводу css.module активной кнопки TabsItem. Хочу поменять стандартный белый активный цвет на кастомный. Пробовал достучаться до него через классы .vkuiTabsItem--selected.vkuiTabsItem--mode-accent .vkuiTabsItem__label разным способом, но что-то так и не выходит. При чём, если меняю свойство на background-color, то всё замечательно работает, как только меняю на color, то активная кнопка меняет свой цвет на стандартный белый, но если же кнопка не активна, то цвет меняется. Все скрины прилагаю ниже. Подскажите, как в css достучаться до цвета активной кнопки TabsItem?

`<TabsItem
            selected={selected === 'all_the_time'}
            onClick={() => setSelected('all_the_time')}
            id="tab-all_the_time"
            aria-controls="tab-content-all_the_time"
            className={selected === 'all_the_time' ? 'TabsText' : ''}>
            За всё время
</TabsItem>`

`:global {
  .TabsText {
    .vkuiTabsItem__label {
      color: aqua;
    }
  }
}`

![1](https://github.com/VKCOM/VKUI/assets/146764865/8bc9b0d4-2c07-4055-b36c-f9ad97112c73)
![2](https://github.com/VKCOM/VKUI/assets/146764865/6816b7ff-caed-45b3-8e5b-8bde137b0ed8)
![3](https://github.com/VKCOM/VKUI/assets/146764865/d8bacf6b-0b4f-4432-909a-98a37560f7a3)
![4](https://github.com/VKCOM/VKUI/assets/146764865/996fd900-7215-4813-8e7d-b36d1d0234e0)
![5](https://github.com/VKCOM/VKUI/assets/146764865/9b7a5e85-04da-4c51-ab7b-ab6f4e44a805)

**Comment by SevereCloud** [comment]
Внутренняя реализация компонентов, включая их классы — не публичное API VKUI и [может](https://ru.wikipedia.org/wiki/%D0%98%D0%BD%D0%BA%D0%B0%D0%BF%D1%81%D1%83%D0%BB%D1%8F%D1%86%D0%B8%D1%8F_(%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)) измениться в любой момент без предупреждения. Не привязывайтесь к ней, если хотите сохранить возможность без проблем обновляться на новые версии библиотеки.

Проблема скорее всего в [специфичности](https://developer.mozilla.org/ru/docs/Web/CSS/Specificity): 
- За неактивный цвет отвечает селектор `.vkuiTabsItem__label`, который имеет специфичность **1**.
- За активный цвет отвечает селектор `.vkuiTabsItem--selected .vkuiTabsItem__label`, который имеет специфичность **2**.
- Селектор `.TabsText .vkuiTabsItem__label` имеет специфичность **2**.

При одинаковой специфичности применяется тот стиль, который находится ниже других. Как вариант можно поменять порядок подключения css или [увеличить](https://developer.mozilla.org/ru/docs/Web/CSS/Specificity#%D0%B2%D0%B0%D0%B6%D0%BD%D0%BE%D0%B5_%D0%B8%D1%81%D0%BA%D0%BB%D1%8E%D1%87%D0%B5%D0%BD%D0%B8%D0%B5_%D0%B8%D0%B7_%D0%BF%D1%80%D0%B0%D0%B2%D0%B8%D0%BB_-_!important) специфичность.

**Reply by Rezetchik** [reply]
Получается, если использовать компоненты VK, то только в том виде, в которым они есть? Как быть, если я хочу кастомизировать компонент/немного видоизменить, например, по цвету и тп? Лучше тогда вообще не использовать компоненты VK и собирать самому? Какие есть способы решения? Не обновляться в библиотеках?

**Reply by SevereCloud** [reply]
> Получается, если использовать компоненты VK, то только в том виде, в которым они есть? 

Верно, компоненты должны соответствовать дизайн системе.

При взаимодействии с компонентами используются его свойства, а все что внутри, это уже реализация. [Инкапсуляция](https://ru.wikipedia.org/wiki/%D0%98%D0%BD%D0%BA%D0%B0%D0%BF%D1%81%D1%83%D0%BB%D1%8F%D1%86%D0%B8%D1%8F_(%D0%BF%D1%80%D0%BE%D0%B3%D1%80%D0%B0%D0%BC%D0%BC%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)) важный механизм в программировании. Его суть в том, что вы разделяете интерфейсы и реализацию. Если вы как-то влияете на реализацию компонентов, вы нарушаете инкапсуляцию.

Инкапсуляция важная часть [веб-компонентов](https://developer.mozilla.org/ru/docs/Web/API/Web_components#%D0%BF%D0%BE%D0%BD%D1%8F%D1%82%D0%B8%D1%8F_%D0%B8_%D0%B8%D1%81%D0%BF%D0%BE%D0%BB%D1%8C%D0%B7%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5:~:text=JavaScript%20%D0%B4%D0%BB%D1%8F%20%D0%BF%D1%80%D0%B8%D0%BA%D1%80%D0%B5%D0%BF%D0%BB%D0%B5%D0%BD%D0%B8%D1%8F-,%D0%B8%D0%BD%D0%BA%D0%B0%D0%BF%D1%81%D1%83%D0%BB%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%BD%D0%BE%D0%B3%D0%BE,-%C2%AB%D1%82%D0%B5%D0%BD%D0%B5%D0%B2%D0%BE%D0%B3%D0%BE%C2%BB%20%D0%B4%D0%B5%D1%80%D0%B5%D0%B2%D0%B0%20DOM). Но из-за некоторых ограничений мы не можем использовать Shadow DOM и вместо этого отдельно отдаем [css modules](https://vkcom.github.io/VKUI/#/CSS%20Modules)

> Как быть, если я хочу кастомизировать компонент/немного видоизменить, например, по цвету и тп?...Какие есть способы решения?

Использовать публичное API компонентов.

Также для некоторых видоизменений можно использовать токены, которые описаны в разделе [кастомизациий](https://vkcom.github.io/VKUI/#/Customize)

**Reply by Rezetchik** [reply]
Где я могу найти публичные API-компоненты ? Как их использовать? Или это и есть те самые свойства и методы компонентов, которые прописаны в https://vkcom.github.io/VKUI/ ?

Прошу рассказать, поподробнее, что значит "на свой страх и риск" про нарушение инкапсуляции. Возможно ли такое, что если я не буду обновлять библиотеку в своём кейсе до полной реализации, деплоита и публикации мини-приложения, то у меня ничего не "слетит". Или же если ваша библиотека обновится до новой версии, то оно внесёт свои изменения в уже опубликованное мини-приложение? Какие последствия грозят нарушения инкапсуляции, привязываясь к классам, учитывая, что в разработке фронта мини-приложения буду участвовать только я и допустим заранее буду знать о проблемах, которые могут возникнуть из-за нарушения инкапсуляции?

Существуют ли какие-то нетривиальные подходы использования компонентов? Например, можно ли полностью подключить компонент со всеми его свойствами и стилями, и их уже видоизменять локально? Или же вся структура компонентов находятся в закрытом доступе? 

На примере данной проблемы (активный цвет текста в компонентах TabsItem), не могли бы вы мне объяснить, как не нарушать инкапсуляцию? Или данный случай предполагает собственную сборку похожего компонента?

**Reply by SevereCloud** [reply]
> Где я могу найти публичные API-компоненты ? Как их использовать? Или это и есть те самые свойства и методы компонентов, которые прописаны в https://vkcom.github.io/VKUI/ ?

Верно, свойства и методы компонентов это публичное API

> Прошу рассказать, поподробнее, что значит "на свой страх и риск" про нарушение инкапсуляции. Возможно ли такое, что если я не буду обновлять библиотеку в своём кейсе до полной реализации, деплоита и публикации мини-приложения, то у меня ничего не "слетит". Или же если ваша библиотека обновится до новой версии, то оно внесёт свои изменения в уже опубликованное мини-приложение?

Библиотека сама не обновится

> Какие последствия грозят нарушения инкапсуляции, привязываясь к классам, учитывая, что в разработке фронта мини-приложения буду участвовать только я и допустим заранее буду знать о проблемах, которые могут возникнуть из-за нарушения инкапсуляции?

Нарушение инкапсуляции грозит тем, что когда вы обновите VKUI, код может перестать работать:
- например у селектора повысится специфичность
- поменяется структура компонента
- переименнуется класс 
- и т.д.

> [!WARNING]
> У завязки на классы есть еще одна особенность, сборщики [не могут гарантировать правильный порядок](https://github.com/webpack/webpack/issues/7094) стилей.

> Существуют ли какие-то нетривиальные подходы использования компонентов? Например, можно ли полностью подключить компонент со всеми его свойствами и стилями, и их уже видоизменять локально? Или же вся структура компонентов находятся в закрытом доступе?

Структура компонентов это реализация.

> На примере данной проблемы (активный цвет текста в компонентах TabsItem), не могли бы вы мне объяснить, как не нарушать инкапсуляцию? 

```css
.colorThis {
  color: aqua;
}
```

```jsx
<TabsItem>
  <span className="colorThis">Text</span>
</TabsItem>
```

**Reply by Rezetchik** [reply]
Спасибо, за ваш ответ! Но остался ещё один вопрос, при таком коде, при переключении между TabsItem, кнопка на пол секунды из color aqua становится белой, а потом уже серой. И получается не совсем красиво :( Пытался найти состояния, отвечающее за "уход" с кнопки, но так и не получилось добиться результата.

Например, при переключение с "Все вкладки" на "Только мои", текст сначала выделяется белым, а потом серым. Нужно именно убрать этот полусекундный белый цвет состояния "ухода с кнопки", или заменять на тот цвет, который мне хочется...

`
<TabsItem>
<span className={selected === 'Только мои' ? 'colorThis' : ' '}> Text </span>
 </TabsItem>
<TabsItem>
<span className={selected === 'Все вкладки' ? 'colorThis' : ' '}> Text </span>
 </TabsItem>
`

`
.colorThis {
  color: aqua;
}
`

**Reply by Rezetchik** [reply]
![1](https://github.com/VKCOM/VKUI/assets/146764865/a95508b1-9f88-4d4e-972c-b2f68572c30e)
![2](https://github.com/VKCOM/VKUI/assets/146764865/d9fc4bd1-dab8-49c8-a87b-e62a0e0fe83d)

**Reply by SevereCloud** [reply, marked as answer]
```css
.selectColor {
  color: aqua;
}
.unselectColor {
  color: var(--vkui--color_text_secondary);
}
```

```jsx
<TabsItem>
  <span className={selected ? 'selectColor' : 'unselectColor'}> Text </span> 
</TabsItem>
```
